embassy-mspm0/src/watchdog.rs

```diff
@@ -0,0 +1,330 @@
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ClosedWindowPercentage {
+    // window period is not used
+    Zero,
+    // 12.5% percents
+    Twelve,
+    // 18.75% percents
+    Eighteen,
+    // 25% percents
+    TwentyFive,
+    // 50% percents
+    Fifty,
+    // 75% percents
+    SeventyFive,
+    // 81.25% percents
+    EightyOne,
+    // 87.5% percents
+    EightySeven,
+}
+
+#[non_exhaustive]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+/// Watchdog Config
+pub struct Config {
+    /// Watchdog timeout
+    pub timeout: Timeout,
+
+    /// closed window percentage
+    pub closed_window: ClosedWindowPercentage,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            timeout: Timeout::Sec1,
+            closed_window: ClosedWindowPercentage::Zero,
+        }
+    }
+}
+
+pub struct Watchdog<'d, T: Instance> {
+    wdt: PhantomData<&'d mut T>,
+}
+
+impl<'d, T: Instance> Watchdog<'d, T> {
+    /// Watchdog initialization.
+    pub fn new(_instance: Peri<'d, T>, config: Config) -> Self {
+        // Init power for watchdog
+        T::REGS.gprcm(0).rstctl().write(|w| {
+            w.set_resetstkyclr(true);
+            w.set_resetassert(true);
+            w.set_key(vals::ResetKey::KEY);
+        });
+
+        // Enable power for watchdog
+        T::REGS.gprcm(0).pwren().write(|w| {
+            w.set_enable(true);
+            w.set_key(vals::PwrenKey::KEY);
+        });
+
+        // init delay, 16 cycles
+        cortex_m::asm::delay(16);
+
+        //init watchdog
+        T::REGS.wwdtctl0().write(|w| {
+            w.set_clkdiv(config.timeout.get_clkdiv());
+            w.set_per(config.timeout.get_period());
+            w.set_mode(vals::Mode::WINDOW);
+            match config.closed_window {
+                ClosedWindowPercentage::Zero => w.set_window0(vals::Window::SIZE_0),
+                ClosedWindowPercentage::Twelve => w.set_window0(vals::Window::SIZE_12),
+                ClosedWindowPercentage::Eighteen => w.set_window0(vals::Window::SIZE_18),
+                ClosedWindowPercentage::TwentyFive => w.set_window0(vals::Window::SIZE_25),
+                ClosedWindowPercentage::Fifty => w.set_window0(vals::Window::SIZE_50),
+                ClosedWindowPercentage::SeventyFive => w.set_window0(vals::Window::SIZE_75),
+                ClosedWindowPercentage::EightyOne => w.set_window0(vals::Window::SIZE_81),
+                ClosedWindowPercentage::EightySeven => w.set_window0(vals::Window::SIZE_87),
+            }
+            w.set_window1(vals::Window::SIZE_0);
+            w.set_key(vals::Wwdtctl0Key::KEY);
+        });
+
+        // Set Window0 as active window
+        T::REGS.wwdtctl1().write(|w| {
+            w.set_winsel(vals::Winsel::WIN0);
+            w.set_key(vals::Wwdtctl1Key::KEY);
+        });
+
+        // make sure watchdog triggers BOOTRST
+        pac::SYSCTL.systemcfg().write(|w| {
```

Comment:
If WWDTLP1RSTDIS is needed then this sysctl access needs to be guarded by a critical section in the case where thread a configures wwdt0 and thread b preempts and configures wwdt1.

Reply:
fixed